pages/blog/subscriptions-and-live-queries-real-time-with-graphql.mdx

```diff
@@ -0,0 +1,867 @@
+
+Usually you will have a network interface that is called by you favorite GraphQL client. Every single one has a different naming and approach of how they work. Apollo calls them links, relay calls the fetcher functions, and urql calls them exchanges.
+
+All of them have one thing in common. They are working with observable like data structures, which basically means all major GraphQL client-libraries decided to use Observables for consuming subscriptions, those in contrast to AsyncIterators are not part of the ECMA Spec (as of October 2020).
+
+Like AsyncIterators a Observable can represent multiple values. As already mentioned client library network interface and transport has a slightly different interface I will use `graphql-ws` with `relay-runtime` as an example.
+
+The example is taken straight from the `graphql-ws` section.
+
+```ts
+import {
+  Network,
+  Observable,
+  RequestParameters,
+  Variables,
+} from 'relay-runtime';
+import { createClient } from 'graphql-ws';
+
+const subscriptionsClient = createClient({
+  url: 'wss://i.love/graphql',
+  connectionParams: () => {
+    const session = getSession();
+    if (!session) {
+      return {};
+    }
+    return {
+      Authorization: `Bearer ${session.token}`,
+    };
+  },
+});
+
+// yes, both fetch AND subscribe handled in one implementation
+function fetchOrSubscribe(operation: RequestParameters, variables: Variables) {
+  return Observable.create((sink) => {
+    if (!operation.text) {
+      return sink.error(new Error('Operation text cannot be empty'));
+    }
+    return subscriptionsClient.subscribe(
+      {
+        operationName: operation.name,
+        query: operation.text,
+        variables,
+      },
+      {
+        ...sink,
+        error: (err) => {
+          if (err instanceof Error) {
+            sink.error(err);
+          } else if (err instanceof CloseEvent) {
+            sink.error(
+              new Error(
+                `Socket closed with event ${err.code}` + err.reason
+                  ? `: ${err.reason}` // reason will be available on clean closes
+                  : ''
+              )
+            );
+          } else {
+            // GraphQLError[]
+            sink.error(new Error(err.map(({ message }) => message).join(', ')));
+          }
+        },
+      }
+    );
+  });
+}
+
+const network = Network.create(fetchOrSubscribe, fetchOrSubscribe);
+const store = new Store(new RecordSource());
+export const environment = new Environment({ network, store });
+```
+
+With this configuration relay can now execute subscription operations. Because the graphql-ws protocol is way more complex than the GraphQL over HTTP protocol, we use the client exported from teh graphql-ws package instead. Some additional bundle-size we got there. As mentioned before SSE might be a better leight-weight alternative.
```

Comment:
```suggestion
With this configuration, relay can now execute subscription operations. Because the graphql-ws protocol is way more complex than the GraphQL over HTTP protocol, we use the client exported from the graphql-ws package instead. Some additional bundle-size we got there. As mentioned before SSE might be a better light-weight alternative.
```